aiida/tools/importexport/dbexport/__init__.py

```diff
@@ -233,193 +226,116 @@ def export(
-        what rule names are toggleable and what the defaults are.
+    # Initialize the writer
+    with writer as writer_context:
 
-    :raises `~aiida.tools.importexport.common.exceptions.ArchiveExportError`:
-        if there are any internal errors when exporting.
-    :raises `~aiida.common.exceptions.LicensingException`:
-        if any node is licensed under forbidden license.
-    """
-    # pylint: disable=too-many-locals
-    EXPORT_LOGGER.debug('STARTING EXPORT...')
+        # Iteratively explore the AiiDA graph to find further nodes that should also be exported
+        with get_progress_reporter()(desc='Traversing provenance via links ...', total=1) as progress:
+            traverse_output = get_nodes_export(starting_pks=given_node_entry_ids, get_links=True, **traversal_rules)
+            progress.update()
+        node_ids_to_be_exported = traverse_output['nodes']
+
+        EXPORT_LOGGER.debug('WRITING METADATA...')
+
+        writer_context.write_metadata(
+            ArchiveMetadata(
+                export_version=EXPORT_VERSION,
+                aiida_version=get_version(),
+                unique_identifiers=unique_identifiers,
+                all_fields_info=all_fields_info,
+                graph_traversal_rules=traverse_output['rules'],
+                # Turn sets into lists to be able to export them as JSON metadata.
+                entities_starting_set={
+                    entity: list(entity_set) for entity, entity_set in entities_starting_set.items()
+                },
+                include_comments=include_comments,
+                include_logs=include_logs,
+            )
+        )
 
-    # Backwards-compatibility
-    entities = cast(
-        Iterable[Any],
-        deprecated_parameters(
-            old={
-                'name': 'what',
-                'value': traversal_rules.pop('what', None)
-            },
-            new={
-                'name': 'entities',
-                'value': entities
-            },
-        ),
-    )
+        # Create a mapping of node PK to UUID.
+        node_pk_2_uuid_mapping: Dict[int, str] = {}
+        if node_ids_to_be_exported:
+            qbuilder = orm.QueryBuilder().append(
+                orm.Node,
+                project=('id', 'uuid'),
+                filters={'id': {
+                    'in': node_ids_to_be_exported
+                }},
+            )
+            node_pk_2_uuid_mapping = dict(qbuilder.all(batch_size=batch_size))
+
+        # check that no nodes are being exported with incorrect licensing
+        _check_node_licenses(node_ids_to_be_exported, allowed_licenses, forbidden_licenses)
+
+        # write the link data
+        if traverse_output['links']:
+            with get_progress_reporter()(total=len(traverse_output['links']), desc='Writing links') as progress:
+                for link in traverse_output['links']:
+                    progress.update()
+                    writer_context.write_link({
```

Comment:
so, this interface seems to assume that links are to be written one at a time (not streamed).
should they not be streamed as well eventually?

Reply:
ok, I see now that also the entities are just written one at a time.
I guess when writing the commit message for this PR, we'll need to be a bit careful with the word "stream" since one could understand something different

Reply:
I'm not sure if I understand this; they are "streamed" to the writer one at a time.
The write can decide how it writes them; caching batches before writing, or otherwise.

Are you suggesting that you should send the links to the writer already in batches?

Reply:
what would you understand by stream?

Reply:
> what would you understand by stream?

On second thought, I think you are right and your use of "stream" makes sense.